src/btree/bt_read.c

```diff
@@ -140,6 +141,242 @@ __instantiate_birthmarks(WT_SESSION_IMPL *session, WT_REF *ref)
+
+        /* Ignore the next set of records for the same key that is already instantiated. */
+        switch (page->type) {
+        case WT_PAGE_COL_FIX:
+        case WT_PAGE_COL_VAR:
+            p = las_key.data;
+            WT_ERR(__wt_vunpack_uint(&p, 0, &recno));
+            if (current_recno == recno) {
+                /*
+                 * For birthmark records insert only the birthmark record, and for other records,
+                 * insert the latest record for that key.
+                 */
+                if (birthmark_record)
+                    continue;
+                else
+                    break;
+            }
+            WT_ASSERT(session, current_recno < recno);
+
+            if (lastupd != NULL) {
+                WT_ERR(__col_instantiate(session, current_recno, ref, &cbt, lastupd));
+
+                /* Remove the prepared record from LAS */
+                if (!birthmark_record && lastupd->prepare_state == WT_PREPARE_INPROGRESS) {
+                    cursor->prev(cursor);
+                    cursor->remove(cursor);
+                    cursor->next(cursor);
+                }
+
+                if (birthmark_record) {
+                    birthmarkp->instantiated = true;
+                    birthmark_record = false;
+                }
+                lastupd = NULL;
+            }
+            current_recno = recno;
+            break;
+        case WT_PAGE_ROW_LEAF:
+            if (current_key->size == las_key.size &&
+              memcmp(current_key->data, las_key.data, las_key.size) == 0) {
+                /*
+                 * For birthmark records insert only the birthmark record, and for other records,
+                 * insert the latest record for that key.
+                 */
+                if (birthmark_record)
+                    continue;
+                else
+                    break;
+            }
+
+            if (lastupd != NULL) {
+                WT_ERR(__row_instantiate(session, current_key, ref, &cbt, lastupd));
+
+                /* Remove the prepared record from LAS */
+                if (!birthmark_record && lastupd->prepare_state == WT_PREPARE_INPROGRESS) {
+                    cursor->prev(cursor);
+                    cursor->remove(cursor);
+                    cursor->next(cursor);
+                }
+
+                if (birthmark_record) {
+                    birthmarkp->instantiated = true;
+                    birthmark_record = false;
+                }
+                lastupd = NULL;
+            }
+            WT_ERR(__wt_buf_set(session, current_key, las_key.data, las_key.size));
+            break;
+        default:
+            WT_ERR(__wt_illegal_value(session, page->type));
+        }
+
+        if (ref->page_las->birthmarks_cnt > 0) {
+            for (; birthmark_cnt < ref->page_las->birthmarks_cnt; birthmark_cnt++) {
+                birthmarkp = ref->page_las->birthmarks + birthmark_cnt;
+                if (birthmarkp->instantiated)
```

Comment:
do you need this check, I don't think you can have any birthmark already instantiated when you loop through the birthmarks here. Can it be replaced by an assert instead?

Reply:
No. When the birthmark record is found for a key, the birthmark pointer is not incremented to next one until we check for the next key, so for the older birthmark record the instantiation flag is set.